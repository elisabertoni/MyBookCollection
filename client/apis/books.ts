import request from 'superagent'
import { Book, BookCreate, BookUpdate } from '../../models/book'

export async function getBooksApi(): Promise<Book[]> {
  const response = await request.get('/api/v1/books')
  return response.body
}

export async function addBookApi(newBook: BookCreate): Promise<Book[]> {
  const response = await request.post('/api/v1/books').send(newBook)
  return response.body
}

export async function updateBookApi(
  bookId: number,
  updatedBook: BookUpdate
): Promise<Book[]> {
  const response = await request
    .patch(`/api/v1/books/${bookId}`)
    .send(updatedBook)
  return response.body
}

export async function deleteBookApi(bookId: number): Promise<Book[]> {
  const response = request.delete(`/api/v1/books/${bookId}`)
  return (await response).body
}
