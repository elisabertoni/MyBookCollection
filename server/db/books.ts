import connection from './connection'
import { Book } from '../../models/book'

// Books functions
export function getAllBooks( db = connection): Promise<Book[]> {
  return db('books').select()
}

export function addBook(newBook: Book, db = connection): Promise<Book[]> {
  return db('books').insert({...newBook})
}

export function updateBook(bookId:number, updatedBook: Book, db = connection): Promise<Book> {
  return db('books')
    .update(updatedBook)
    .where('id', bookId )
    .returning(['id', 'title', 'author'])
    .then((result) => result[0])
}