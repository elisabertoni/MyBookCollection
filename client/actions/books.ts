import { ThunkAction } from "../store";
import { getBooksApi, addBookApi, updateBookApi, deleteBookApi } from "../apis/books";
import { Book, BookCreate } from "../../models/book";

export const SET_BOOK_PENDING = 'SET_BOOKS_PENDING'
export const SET_BOOK_SUCCESS = 'SET_BOOKS_SUCCESS'
export const SET_BOOK_ADD = 'SET_BOOK_ADD'
export const SET_BOOK_UPDATE = 'SET_BOOK_UPDATE'
export const SET_BOOK_DELETE = 'SET_BOOK_DELETE'
export const SET_ERROR = 'SET_PENDING'

export type BookAction = 
  | { type: typeof SET_BOOK_PENDING; payload: null }
  | { type: typeof SET_BOOK_SUCCESS ; payload: Book[] }
  | { type: typeof SET_BOOK_ADD ; payload: Book[] }
  | { type: typeof SET_BOOK_UPDATE ; payload: Book }
  | { type: typeof SET_BOOK_DELETE ; payload: null }
  | { type: typeof SET_ERROR; payload: string }


export function setBooksPending(): BookAction {
  return {
    type: SET_BOOK_PENDING,
    payload: null 
  }
}

export function setBooksSuccess(books: Book[]): BookAction {
  return {
    type: SET_BOOK_SUCCESS,
    payload: books
  }
}

export function setBookAdd(books: Book[]): BookAction {
  return {
    type: SET_BOOK_ADD,
    payload: books
  }
}

export function setBookUpdate(books: Book): BookAction {
  return {
    type: SET_BOOK_UPDATE,
    payload: books
  }
}

export function setBookDelete(): BookAction {
  return {
    type: SET_BOOK_DELETE,
    payload: null
  }
}
export function setError(errMessage: string): BookAction {
  return {
    type: SET_ERROR,
    payload: errMessage
  }
}

export function fetchBooks(): ThunkAction {
  return dispatch => {
    dispatch(setBooksPending())
    return getBooksApi()
    .then(books => {
      dispatch(setBooksSuccess(books))
    })
    .catch(err => {
      dispatch(setError(err.message))
    })
  }
}

export function addBook(newBook: BookCreate): ThunkAction {
  return dispatch => {
    return addBookApi(newBook)
    .then(books => {
      dispatch(setBookAdd(books))
    })
    .catch(err => {
      dispatch(setError(err.message))
    })
  }
}

export function updateBook(bookId:number, book: Book): ThunkAction {
  return dispatch => {
    return updateBookApi(bookId, book)
    .then(book => {
      dispatch(setBookUpdate(book))
      dispatch(fetchBooks())
    })
    .catch(err => {
      dispatch(setError(err.message))
    })
  }
}

export function deleteBook(bookId: number): ThunkAction {
  return dispatch => {
    return deleteBookApi(bookId)
    .then(() => {
      dispatch(setBookDelete())
      dispatch(fetchBooks())
    })
    .catch(err => {
      dispatch(setError(err.message))
    })
  }
}