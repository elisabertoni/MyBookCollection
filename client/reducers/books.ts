import { SET_BOOK_PENDING, SET_BOOK_SUCCESS, SET_ERROR, BookAction } from "../actions/books";
import { Book } from "../../models/book";

interface BookState {
  loading: boolean
  data: Book[]
  error: string | undefined
}

const initialState: BookState = {
  loading: false,
  data: [],
  error: undefined
}

export default function reducer(state = initialState, action: BookAction): BookState {
  switch (action.type) {
    case SET_BOOK_PENDING:
      return {
        loading: true,
        data: [],
        error: undefined
      }
    case SET_BOOK_SUCCESS:
      return {
        loading: false,
        data: action.payload,
        error: undefined
      }
    case SET_ERROR:
      return {
        loading: false,
        data: [],
        error: action.payload
      }
    default:
      return state
  }
}