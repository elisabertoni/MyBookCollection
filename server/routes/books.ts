import express from 'express'
import { addBook, deleteBook, getAllBooks, updateBook } from '../db/books'
const router = express.Router()

//GET /api/v1/books
router.get('/', async (req, res) => {
  try {
    const books = await getAllBooks()
    res.json(books)

  } catch (error) {
    console.log(error)
    res.status(500).json({
      error: "There was an error trying to get the books"
    })
  }
})

// POST /api/v1/books
router.post('/', async (req, res) => {
  try {
    const newBook = req.body

    const [newBookId] = await addBook(newBook)

    res.json({ id: newBookId, ...newBook })
    
  } catch (error) {
    console.log(error)
    res.status(500).json({
      error: "There was an error trying to add the book"
    })
  }
})

// PATCH  /api/v1/books/:id
router.patch('/:id', async (req, res) => {
  try {
    const bookId = req.params.id

    if (isNaN(Number(bookId))) {
      res.status(400).json({
        error: 'Invalid Book ID',
      })
    }

    const updatedBook= await updateBook(Number(bookId), req.body)
    res.json(updatedBook)

  } catch (error) {
    console.log(error)
    res.status(500).json({
      error: "There was an error trying to update the book"
    })
  }
})

// DELETE  /api/v1/books/:id
router.delete('/:id', async (req, res) => {
  try {
    const bookId = req.params.id

    if (isNaN(Number(bookId))) {
      res.status(400).json({
        error: 'Invalid Book ID',
      })
    }

    await deleteBook(Number(bookId))
    res.status(200).json('OK')

  } catch (error) {
    console.log(error)
    res.status(500).json({
      error: "There was an error trying to delete the book"
    })
  }
})

export default router