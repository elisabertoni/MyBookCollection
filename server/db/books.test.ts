import connection from './connection'
import { getAllBooks, addBook, updateBook, deleteBook } from '../db/books'

beforeAll(() => {
  return connection.migrate.latest()
})

beforeEach(() => {
  return connection.seed.run()
})

afterAll(() => {
  return connection.destroy()
})

describe('database tests', () => {
  it('getAllBooks works', async () => {
    // Arrange
    // Act
    const books = await getAllBooks()
    // Assert
    expect(books).toMatchSnapshot(`[
      { id: 1, title: 'The Catcher in the Rye', author: 'J.D.Salinger' },
      { id: 2, title: 'It Ends with Us', author: 'Colleen Hoover' },
      {
        id: 3,
        title: "Harry Potter and the Philosopher's Stone",
        author: 'J.K.Rowling'
      },
      {
        id: 4,
        title: 'The Lion, the Witch, and the Wardrobe',
        author: 'C.S. Lewis'
      },
      { id: 5, title: 'Normal People', author: 'Sally Rooney' },
      { id: 6, title: 'The Duke and I', author: 'Julia Quinn' },
      { id: 7, title: 'Girls That Invest', author: 'Simran Kaur' },
      {
        id: 8,
        title: 'My Darling Lemon Thyme: Every Day',
        author: 'Emma Galloway'
      },
      { id: 9, title: 'Search Inside Yourself', author: ' Chade-Meng Ta' },
      {
        id: 10,
        title: 'Homegrown Kitchen: Everyday Recipes for Eating Well',
        author: 'Nicola Galloway'
      },
      {
        id: 11,
        title: 'Healthy Baking: Nourishing breads, wholesome cakes, ancient grains and bubbling ferments',
        author: 'Jordan Bourke'
      }
    ]`)
  })
  it('add a new book', async () => {
    
  })
})
