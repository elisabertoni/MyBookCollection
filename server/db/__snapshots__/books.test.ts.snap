// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`database tests getAllBooks works: [
      { id: 1, title: 'The Catcher in the Rye', author: 'J.D.Salinger' },
      { id: 2, title: 'It Ends with Us', author: 'Colleen Hoover' },
      {
        id: 3,
        title: "Harry Potter and the Philosopher's Stone",
        author: 'J.K.Rowling'
      },
      {
        id: 4,
        title: 'The Lion, the Witch, and the Wardrobe',
        author: 'C.S. Lewis'
      },
      { id: 5, title: 'Normal People', author: 'Sally Rooney' },
      { id: 6, title: 'The Duke and I', author: 'Julia Quinn' },
      { id: 7, title: 'Girls That Invest', author: 'Simran Kaur' },
      {
        id: 8,
        title: 'My Darling Lemon Thyme: Every Day',
        author: 'Emma Galloway'
      },
      { id: 9, title: 'Search Inside Yourself', author: ' Chade-Meng Ta' },
      {
        id: 10,
        title: 'Homegrown Kitchen: Everyday Recipes for Eating Well',
        author: 'Nicola Galloway'
      },
      {
        id: 11,
        title: 'Healthy Baking: Nourishing breads, wholesome cakes, ancient grains and bubbling ferments',
        author: 'Jordan Bourke'
      }
    ] 1`] = `
[
  {
    "author": "J.D.Salinger",
    "id": 1,
    "title": "The Catcher in the Rye",
  },
  {
    "author": "Colleen Hoover",
    "id": 2,
    "title": "It Ends with Us",
  },
  {
    "author": "J.K.Rowling",
    "id": 3,
    "title": "Harry Potter and the Philosopher's Stone",
  },
  {
    "author": "C.S. Lewis",
    "id": 4,
    "title": "The Lion, the Witch, and the Wardrobe",
  },
  {
    "author": "Sally Rooney",
    "id": 5,
    "title": "Normal People",
  },
  {
    "author": "Julia Quinn",
    "id": 6,
    "title": "The Duke and I",
  },
  {
    "author": "Simran Kaur",
    "id": 7,
    "title": "Girls That Invest",
  },
  {
    "author": "Emma Galloway",
    "id": 8,
    "title": "My Darling Lemon Thyme: Every Day",
  },
  {
    "author": " Chade-Meng Ta",
    "id": 9,
    "title": "Search Inside Yourself",
  },
  {
    "author": "Nicola Galloway",
    "id": 10,
    "title": "Homegrown Kitchen: Everyday Recipes for Eating Well",
  },
  {
    "author": "Jordan Bourke",
    "id": 11,
    "title": "Healthy Baking: Nourishing breads, wholesome cakes, ancient grains and bubbling ferments",
  },
]
`;
